src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,972 @@
+            {
+                var symbol = semanticModel.GetSymbolInfo(variable, cancellationToken).Symbol;
+                if (symbol is IParameterSymbol parameterSymbol)
+                {
+                    nameToParameterDict.Add(variable, parameterSymbol);
+                }
+            }
+
+            return nameToParameterDict;
+        }
+
+        /// <summary>
+        /// Determines if the expression is something that should have code actions displayed for it.
+        /// Depends upon the identifiers in the expression mapping back to parameters.
+        /// Does not handle params parameters.
+        /// </summary>
+        private async Task<(bool shouldDisplay, bool containsClassExpression)> ShouldExpressionDisplayCodeActionAsync(
+            Document document, TExpressionSyntax expression,
+            CancellationToken cancellationToken)
+        {
+            var variablesInExpression = expression.DescendantNodes();
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var containsClassSpecificStatement = false;
+
+            foreach (var variable in variablesInExpression)
+            {
+                var symbol = semanticModel.GetSymbolInfo(variable, cancellationToken).Symbol;
+                if (symbol is IRangeVariableSymbol or ILocalSymbol)
+                {
+                    return (false, false);
+                }
+
+                if (symbol is IParameterSymbol parameter)
+                {
+                    if (parameter.IsParams)
+                    {
+                        return (false, false);
+                    }
+                }
+
+                if (IsClassSpecificExpression(variable))
+                {
+                    containsClassSpecificStatement = true;
+                }
+            }
+
+            return (true, containsClassSpecificStatement);
+        }
+
+        /// <summary>
+        /// Creates new code actions for each introduce parameter possibility.
+        /// Does not create actions for overloads/trampoline if there are optional parameters or if the methodSymbol
+        /// is a constructor.
+        /// </summary>
+        private async Task<(ImmutableArray<CodeAction> actions, ImmutableArray<CodeAction> actionsAllOccurrences)?> GetActionsAsync(Document document,
+            TExpressionSyntax expression, IMethodSymbol methodSymbol, SyntaxNode containingMethod,
+            CancellationToken cancellationToken)
+        {
+            var (shouldDisplay, containsClassExpression) = await ShouldExpressionDisplayCodeActionAsync(
+                document, expression, cancellationToken).ConfigureAwait(false);
+            if (!shouldDisplay)
+            {
+                return null;
+            }
+
+            var actionsBuilder = ImmutableArray.CreateBuilder<CodeAction>();
+            var actionsBuilderAllOccurrences = ImmutableArray.CreateBuilder<CodeAction>();
```

Comment:
use either ArrayBuidler or TemporaryArray.  the latter is actually preferred here as we have a very small amount of these items.

Reply:
swqitched